src/gui/tray/UnifiedSearchResultsListModel.cpp

```diff
@@ -0,0 +1,675 @@
+/*
+ * Copyright (C) by Oleksandr Zolotov <alex@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "UnifiedSearchResultsListModel.h"
+
+#include "account.h"
+#include "accountstate.h"
+#include "guiutility.h"
+#include "folderman.h"
+#include "networkjobs.h"
+
+#include <algorithm>
+
+#include <QAbstractListModel>
+#include <QDesktopServices>
+
+namespace {
+static QString imagePlaceholderUrlForProviderId(const QString &providerId)
+{
+    if (providerId.contains(QStringLiteral("message"), Qt::CaseInsensitive)
+        || providerId.contains(QStringLiteral("talk"), Qt::CaseInsensitive)) {
+        return QStringLiteral("qrc:///client/theme/black/wizard-talk.svg");
+    } else if (providerId.contains(QStringLiteral("file"), Qt::CaseInsensitive)) {
+        return QStringLiteral("qrc:///client/theme/black/edit.svg");
+    } else if (providerId.contains(QStringLiteral("deck"), Qt::CaseInsensitive)) {
+        return QStringLiteral("qrc:///client/theme/black/deck.svg");
+    } else if (providerId.contains(QStringLiteral("calendar"), Qt::CaseInsensitive)) {
+        return QStringLiteral("qrc:///client/theme/black/calendar.svg");
+    } else if (providerId.contains(QStringLiteral("mail"), Qt::CaseInsensitive)) {
+        return QStringLiteral("qrc:///client/theme/black/email.svg");
+    } else if (providerId.contains(QStringLiteral("comment"), Qt::CaseInsensitive)) {
+        return QStringLiteral("qrc:///client/theme/black/comment.svg");
+    }
+
+    return QStringLiteral("qrc:///client/theme/change.svg");
+}
+
+static QString iconUrlForDefaultIconName(const QString &defaultIconName)
+{
+    const QUrl urlForIcon(defaultIconName);
+
+    if (urlForIcon.isValid() && !urlForIcon.scheme().isEmpty()) {
+        return defaultIconName;
+    }
+    
+    if (defaultIconName.startsWith(QStringLiteral("icon-"))) {
+        const auto parts = defaultIconName.split(QLatin1Char('-'));
+
+        if (parts.size() > 1) {
+            const QString iconFilePath = QStringLiteral(":/client/theme/") + parts[1] + QStringLiteral(".svg");
+
+            if (QFile(iconFilePath).exists()) {
```

Comment:
You can use the static function https://doc.qt.io/qt-5/qfile.html#exists It should have better performance.